### ISSUE TITLE
Last part of multi-upload is ignored 

### ISSUE THREAD

**Issue by ringge** [tus/tus-node-server#415]
Hello, it seems the issue of "Last part of multi-upload is ignored" (https://github.com/tus/tusd/pull/462) is not yet fixed in tus/server and @tus/s3store, so now when I integrate @tus/server and @tus/s3store with Nodejs (specifically nextjs) I have the issue that last part of the uploads were ignored, thus the uploaded file is incomplete

**Comment by Murderlon** [comment]
Hi, I haven't seen this happen. We also have a test for this that runs against all stores:

https://github.com/tus/tus-node-server/blob/a89eaa2ae070f3e7a47f51e10fd8f0934e8cda8b/test/stores.test.ts#L191-L214

If you could provide a failing test that would help a lot 🙌

**Comment by ringge** [comment]
> Hi, I haven't seen this happen. We also have a test for this that runs against all stores:
> 
> https://github.com/tus/tus-node-server/blob/a89eaa2ae070f3e7a47f51e10fd8f0934e8cda8b/test/stores.test.ts#L191-L214
> 
> If you could provide a failing test that would help a lot 🙌

Hi, thanks for prompt reply, I will try to provide a failing test then

**Comment by Murderlon** [comment]
@ringge are you still interested in contributing?

**Comment by yalsayid** [comment]
I've encountered the same issue. In my production environment, files are getting mostly uploaded, but the final part is being stored separately, as you can see in the attached screenshots.

<img width="1195" alt="Screenshot 2023-06-22 at 11 17 27 AM" src="https://github.com/tus/tus-node-server/assets/15058134/334551e5-65db-4c17-8b6c-5ea37c716725">
<img width="955" alt="Screenshot 2023-06-22 at 11 13 37 AM" src="https://github.com/tus/tus-node-server/assets/15058134/7bf55793-c432-4c0a-8262-26f9700d8cea">

What seems to have fixed it is setting the chunk part size to 6mb on both the client and server, it's something I noticed in the [Supabase docs](https://supabase.com/docs/guides/storage/uploads#resumable-upload) where they write ```chunkSize: 6 * 1024 * 1024, // NOTE: it must be set to 6MB (for now) do not change it``` and I know they use tus-node-server so I figured I'd try it.

I'm currently working to replicate a failing test, while familiarizing myself with the code base, but right now I'm not sure what the issue is.

**Comment by Murderlon** [comment]
I'll look into it soon.

> setting the chunk part size to 6mb on both the client and server

Note that you should [never](https://github.com/tus/tus-js-client/blob/main/docs/api.md#chunksize) set the `chunkSize` on the client unless your are forced to due to restrictions in your setup. It's always more efficient to upload a file over a single connection instead of multiple. Setting the client chunk size also doesn't impact how chunks are uploaded to S3.

> I'm currently working to replicate a failing test,

Thanks for the effort :)

**Comment by Murderlon** [comment]
> I know they use tus-node-server

Also curious how you know this 😯

**Comment by ringge** [comment]
> 

they said in in Supabase docs that they use Tus protocol : ) https://supabase.com/docs/guides/storage/uploads#resumable-upload

**Comment by Murderlon** [comment]
I saw that! But that could be one of dozens of tus server implementations, I was curious if I missed something specifically about tus-node-server.

**Comment by yalsayid** [comment]
They specifically call out tus-node-server in their [announcement](https://supabase.com/blog/storage-v3-resumable-uploads#implementing-tus-inside-supabase-storage:~:text=To%20implement%20the,after%20Launch%20Week.) for resumable uploads. It appears that they plan to contribute some improvements, which would be awesome!

**Comment by yalsayid** [comment]
> Note that you should [never](https://github.com/tus/tus-js-client/blob/main/docs/api.md#chunksize) set the `chunkSize` on the client unless your are forced to due to restrictions in your setup. It's always more efficient to upload a file over a single connection instead of multiple. Setting the client chunk size also doesn't impact how chunks are uploaded to S3.

Thanks for the heads up, so maybe it was just changing the chunk size in the s3 store config that helped.

**Comment by fenos** [comment]
I came across this issue as well and thought that the cause could have been this issue #444 which has been addressed.
If any of you still encounter this issue after using the latest version please let us know on this thread

**Comment by yalsayid** [comment]
Yeah I am continuing to experience the issue, even with the latest version. I've tried debugging the problem but unfortunately came up empty handed. However, I've found a temporary solution by adjusting the 'chunkSize' within the tus-js-client to 6mb, which fixes the issue. Although @Murderlon previously recommended not to, it definitely fixes the issue for me, but I can't explain why.

I've also attempted to replicate the issue through a failing test but haven't been able to get consistent results. If I modify the test file size in the S3 data store test, it occasionally triggers an error. Here's what I've been using: this.testFileSize = 20_849_384; this.testFileName = '[test.stl](https://drive.google.com/file/d/15afriXhEvbIAYI2SIMdAbQPS8brjDdWK/view?usp=sharing)'.

**Comment by Murderlon** [comment]
Alright, thanks for reporting back.

**Comment by elliotdickison** [comment]
@Murderlon I patched in #475 but it unfortunately it didn't solve the issue for me - I'm still getting corrupted uploads that are missing bytes from the end. Setting `partSize: 6 * 1024 * 1024` seems to fix the issue. I'm using tus & s3-store 1.0.0-beta.7 and uploading to CloudFlare R2. Here's a sample file that triggers the issue 100% of the time for me: https://samplelib.com/lib/preview/mp4/sample-15s.mp4

**Comment by Murderlon** [comment]
Thanks for immediately testing this yourself! I'll take a look next week and update the PR 👍

**Comment by Murderlon** [comment]
Found the issue and incorporated it in the PR.

**Comment by elliotdickison** [comment]
Excellent, thank you so much!

**Comment by mitjap** [comment]
This issue is still present with latest fix. Problem is that all logic regarding current part number, offset or determining whether this is final part must be synchronous. Failing to do so can cause parts to be reordered or not properly detected as final (more common case).

**Comment by elliotdickison** [comment]
@mitjap this seems to have resolved a bunch of upload problems our users were having - thank you!
